File: problems/binary-search/0074-search-a-2d-matrix.js
/**
 * #74 Search a 2D Matrix
 * Pattern: binary search over a virtual one-dimensional sorted array.
 * Map an index to row = floor(index / columns), column = index % columns.
 * Time: O(log(mn)) | Space: O(1)
 */

/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function (matrix, target) {
  if (matrix.length === 0 || matrix[0].length === 0) return false

  const rows = matrix.length
  const columns = matrix[0].length
  let left = 0
  let right = rows * columns - 1

  while (left <= right) {
    const middle = left + Math.floor((right - left) / 2)
    const value = matrix[Math.floor(middle / columns)][middle % columns]

    if (value === target) return true
    if (value < target) left = middle + 1
    else right = middle - 1
  }

  return false
}

module.exports = searchMatrix

File: problems/binary-search/0074-search-a-2d-matrix.py
"""#74 Search a 2D Matrix
Binary-search the matrix as a virtual sorted one-dimensional array.
Time: O(log(mn)) | Space: O(1)
"""


class Solution:
    def searchMatrix(self, matrix, target):
        if not matrix or not matrix[0]:
            return False

        rows, columns = len(matrix), len(matrix[0])
        left, right = 0, rows * columns - 1

        while left <= right:
            middle = left + (right - left) // 2
            value = matrix[middle // columns][middle % columns]
            if value == target:
                return True
            if value < target:
                left = middle + 1
            else:
                right = middle - 1

        return False
